src/xmpp/xmpp-im/xmpp_forwarding.cpp

```diff
@@ -0,0 +1,233 @@
+/*
+ * xmpp_forwarding.cpp - Stanza Forwarding (XEP-0297)
+ * Copyright (C) 2019  Aleksey Andreev
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+#include "xmpp_forwarding.h"
+#include "xmpp_xmlcommon.h"
+#include "xmpp_tasks.h"
+#include "xmpp_client.h"
+#include "xmpp_stream.h"
+#include "xmpp_message.h"
+
+using namespace XMPP;
+
+static const QLatin1String xmlns_forward("urn:xmpp:forward:0");
+static const QLatin1String xmlns_delay("urn:xmpp:delay");
+
+//--------------------------------------------------
+// class Forwarding
+//--------------------------------------------------
+
+Forwarding::Forwarding()
+    : type_(ForwardedNone)
+{
+}
+
+Forwarding::Forwarding(const Forwarding &other)
+    : type_(other.type_)
+    , ts_(other.ts_)
+{
+    if (other.msg_.get())
+        msg_.reset(new Message(*other.msg_.get()));
+}
+
+Forwarding::~Forwarding()
+{
+}
+
+Forwarding & Forwarding::operator=(const Forwarding &from)
+{
+    type_ = from.type_;
+    ts_ = from.ts_;
+    if (from.msg_.get())
+        msg_.reset(new Message(*from.msg_.get()));
+    else
+        msg_.reset(nullptr);
+    return *this;
+}
+
+Forwarding::Type Forwarding::type() const
+{
+    return type_;
+}
+void Forwarding::setType(Type type)
+{
+    if (type_ != type) {
+        type_ = type;
+        if (type == ForwardedNone) {
+            ts_ = QDateTime();
+            ts_.isNull();
+            msg_.reset(nullptr);
+        }
+    }
+}
+
+bool Forwarding::isCarbons() const
+{
+    return (type_ == ForwardedCarbonsSent || type_ == ForwardedCarbonsReceived);
+}
+
+QDateTime Forwarding::timeStamp() const
+{
+    if (!ts_.isNull())
+        return ts_;
+    return msg_.get() ? msg_->timeStamp() : QDateTime();
+}
+
+void Forwarding::setTimeStamp(QDateTime ts)
+{
+    ts_ = ts;
+}
+
+Message *Forwarding::message() const
+{
+    return msg_.get();
+}
+
+void Forwarding::setMessage(const Message &msg)
+{
+    msg_.reset(new Message(msg));
```

Comment:
all this mixing of smart and regular pointers look so weird. I believe if would be much better to introduce NULL Message concept when Private data is not allocated for the message.
Then regular assignment `msg_ = msg` will be enough.
I'll add relevant changes to the Message class.

Reply:
pushed